README.md
Simplecpp-on-windows
====================

Tool to learn programming


History.txt
30/02/14  Using WinBGIm library in place of Xlib
04/03/14  Working except events, text colours
11/03/14  Get click, next event is added
26/03/14  simplecpp library created
29/03/14  Naming conflict solved, Rectangle Polygon working	
30/03/14  Resolved all naming conflicts Rectangle,Polygon,COLOR
01/03/14  Setup file created WinBGIm wizardnot displayed on Codeblocks
03/03/14  Both winbgim and simplecpp is displayed on Codeblocks
05/04/14  verify old version of simplecpp and ask for uninstallation
12/04/14  Integrated Simplecpp on Codeblocks-EP
15/04/14  Mouse events implemented 
16/04/14  Key functions added
22/04/14  Memory leaks bug are fixed.



src/sprite.cpp
#include <simplecpp>
#include <graphics.h>
float Sprite::max_z_index;

Sprite::Sprite(Composite* owner) {
  init(owner);
}

Sprite::Sprite(double dx, double dy, Composite* owner)
  : pose(dx,dy) {
  init(owner);
}
void Sprite::reset(double dx, double dy, Composite* owner){
  pose = Pose(dx,dy);
  repaint();
}
Sprite& Sprite::operator=(const Sprite& other){
  hide();
  z_index = other.z_index; // should z_index change?  also on copy const.
  pose = other.pose;
  visible = other.visible;
  color = other.color;
  fill = other.fill;
  penIsDown = other.penIsDown;
  return *this;
}


Sprite::Sprite(const Sprite& other):
  z_index(other.z_index), pose(other.pose), visible(other.visible),
  color(other.color), fill(other.fill), penIsDown(other.penIsDown)
{
  //if(owner != NULL) owner->addPart(this);
  //else               // will not work with Composite.
  addSprite(this);     // Register to canvas
}

void Sprite::init(Composite* owner){
  visible = true;
  //color = COLOR("black");
  color = COLOR(0,0,0);
  fill = false;
  penIsDown = false;
  z_index = ++ max_z_index;
  if(owner != NULL) owner->addPart(this);
  else addSprite(this);     // Register to canvas
}

Sprite::~Sprite(){
  removeSprite(this);
}

void Sprite::show(){
  visible = true;
  repaint();
}

void Sprite::hide(){
  visible = false;
  repaint();
}

void Sprite::rotate(double angle, bool repaintP){
  pose = Pose(pose, Pose(angle));  // rotate in its frame of reference
  if(repaintP) repaint();
}

void Sprite::setScale(double factor, bool repaintP){
  pose.setScale(factor);
  if(repaintP) repaint();
}

void Sprite::scale(double factor, bool repaintP){
  pose = Pose(pose,Pose(factor,true));
  if(repaintP) repaint();
}

void Sprite::moveTo(double x, double y, bool repaintP){
  Position oldorigin = pose.getOrigin();
  move(x-oldorigin.getX(), y-oldorigin.getY(), repaintP);
}

void Sprite::move(double xshift, double yshift, bool repaintP){
  Position oldorigin = pose.getOrigin();
  pose = Pose(Pose(xshift,yshift),pose);  // move in parent's frame

  if(penIsDown){
    Line l(oldorigin.x, oldorigin.y,
	 oldorigin.x+xshift, oldorigin.y+yshift);
    //l.setColor(COLOR("green"));
    l.imprint(false);
  }
  if(repaintP) repaint();
}

void Sprite::setFill(bool v, bool repaintP){
  fill = v;
  if(repaintP) repaint();
}

Sprite & Sprite::setColor(Color c, bool repaintP){
  color = c;
  if(repaintP) repaint();
  return *this;
}


void Sprite::imprint(bool repaintP){
  c_imprint(this);
  if(repaintP) repaint();
}

void Sprite::forward(double distance){
  Position image  = pose.unitX();
  double xshift = image.getX()*distance;
  double yshift = image.getY()*distance;
  move(xshift,yshift);
}

void Sprite::penDown(bool down){
  penIsDown = down;
}

void Sprite::penUp(bool up){
  penIsDown = !up;
}

void Sprite::face(double x, double y, bool repaintP){
  Position image  = pose.unitX();
  double angle = atan2(image.getY(), image.getX());
  Position origin = getOrigin();
  double target = atan2(y-origin.getY(), x-origin.getX());
  rotate((target - angle));
}
void Sprite::face(Sprite &s){
  Position org = s.getOrigin();
  double x = org.getX();
  double y = org.getY();
  face(x,y);
}

void Sprite::setZIndex(float new_z_index){
  beginFrame();
  removeSprite(this);
  z_index = new_z_index;
  addSprite(this);
  endFrame();
}

float Sprite::getZIndex() const{
  return z_index;
}

Position Sprite::getAtan2() const{
  return pose.getAtan2();
}


src/polygon.cpp
#include <simplecpp>


Polygon::Polygon(Composite* owner){
  init(canvas_width()/2, canvas_height()/2, NULL, 0, owner);
}

Polygon::Polygon(const double x, const double y, const double points[][2],
		 int no, Composite *owner){
  init(x, y, points,no, owner);
}

void Polygon::reset(const double x, const double y,
		    const double points[][2], int no, Composite *owner){
  vertex.clear();
  for(int iter = 0; iter < no; iter ++){
    vertex.push_back(Position(points[iter][0], points[iter][1]));
  }
  Sprite::reset(x,y,owner);
}

void Polygon::init(const double x, const double y,
		   const double points[][2], int no, Composite *owner){
  reset(x, y, points, no, owner);
  show();
}


void Polygon::paint(Pose *p){
  if(visible && vertex.size()>0){
    XPoint xpt_vertex[vertex.size()];
    if(p){
      Pose r = Pose(*p,pose);
      r.compute_res_vertex(xpt_vertex, vertex, vertex.size());
    }
    else pose.compute_res_vertex(xpt_vertex, vertex, vertex.size());
    drawPolygon(xpt_vertex, vertex.size(), color, fill);
  }
}


include/canvas.h
#ifndef _CANVAS_INCLUDED
#define _CANVAS_INCLUDED

//#include <simplecpp>
#include <graphics.h>
namespace canvas{
  double randuv(double u, double v);
  void srand(unsigned int);
  void wait(float duration);
  void abort();

  int initCanvas(const char window_title[]="Simplecpp Canvas",
		 int w=500, int h=500);
  // creates the graphics window

  void closeCanvas();

  int canvas_width();  // return the width of the canvas
  int canvas_height();
  int textWidth(string text);
  int textWidth(char op);
  int textHeight();
  int textDescent();

  void addSprite(Sprite *t);  // adds *t into the list of active sprites
                              // does not repaint immediately?
  void removeSprite(Sprite *t);



  void imprintLine(short x1, short y1, short x2, short y2,
		   Color line_color=COLOR(0,0,0), unsigned int line_width=2);
  // paints the line into the the background

  /* draw... : draw ... onto the current buffer. */

  void drawLine(XPoint start, XPoint end, Color line_color, unsigned int line_width=2);

  void drawPoint(XPoint point, Color point_color);

  void drawCircle(XPoint centre, int radius, Color fill_color, bool fill=true, unsigned int line_width=2, int line_style=SOLID_LINE); 

  void drawEllipse(XPoint centre, int width, int height, Color fill_color, bool fill=true, unsigned int line_width=2, int line_style=SOLID_LINE); 

  void drawPolygon(XPoint *points, int npoints, Color fill_color, bool fill=true, unsigned int line_width=2, int line_style=SOLID_LINE); 

  void drawText(XPoint position, const char *text, Color clr);
  void drawText(Position position, string text, Color clr = COLOR(0,0,255)); 
  void drawText(float x, float y, string text, Color clr = COLOR(0,0,255));
  void drawText(XPoint position, string message, Color clr);

  void c_imprint(Sprite* s);  // paints *s into the background
  void repaint();  // paint the current screen into new buffer, then transfer.
  void beginFrame();  // suspend painting.  set flag appropriately.
  void endFrame();    // resume painting.  reset flag.

  void nextEvent(XEvent &event);  // wait for an event: mouseclick, keypress.
bool mouseDragEvent(XEvent &event);  // true if event is drag
bool keyPressEvent(XEvent &event);
bool mouseButtonPressEvent(XEvent &event);
bool mouseButtonReleaseEvent(XEvent &event);
char charFromEvent(XEvent &event);  // return char whose key was pressed.
 bool checkEvent(XEvent &event);   // true if some event happened since last nextEvent.
void echoKey(XEvent &event);  // print value on screen
  void spriteStatus();
int getClick();  // wait for a click, and return xclickpos*65536+yclickpos.
}
Color COLOR(const char *color_string);  // creates a colour.
  //Color COLOR(unsigned short red, unsigned short green, unsigned short blue);
#endif


src/rectangle.cpp
#include <simplecpp>

Rectangle::Rectangle(Composite* owner)
  : Polygon(0,0,NULL,0,owner){
  init(0,0,0,0);
}

Rectangle::Rectangle(double cx, double cy, double w, double h, Composite* owner)
  : Polygon(cx, cy, NULL, 0, owner){
  init(cx,cy,w,h);
}

void Rectangle::init(double cx, double cy, double w, double h, Composite* owner){
  reset(cx,cy,w,h, owner);
}

void Rectangle::reset(double cx, double cy, double w, double h, Composite* owner){
  //  pose = Pose(cx,cy);

  if(w == -1) w = 10;
  if(h == -1) h = w;
  width = w;
  height = h;

  double points[4][2];
  double hw = w/2, hh = h/2;
  points[0][0] = - hw;
  points[0][1] = - hh;
  points[1][0] = - hw;
  points[1][1] = + hh;
  points[2][0] = + hw;
  points[2][1] = + hh;
  points[3][0] = + hw;
  points[3][1] = - hh;

  Polygon::reset(cx, cy, points,4, owner);
}

double Rectangle::getWidth(){ return width; }
double Rectangle::getHeight(){ return height; }





src/turtleSim.cpp
#include <simplecpp>
#include <cstdlib>

Turtle *_PsystemTurtle_;

void check(const char* msg){
  if(!_PsystemTurtle_){
    cout << "You need to call turtleSim before calling "<<msg<<".\n";
    
    exit(1);
  }
}


namespace tSim{

  void turtlesim(int w, int h){
    turtleSim(w,h);
  }
  void turtleSim(int w, int h){
    initCanvas("Turtle Simulator",w,h);
    _PsystemTurtle_=new Turtle;
    
    
  }

  void closeTurtleSim(){
    closeCanvas();
  }

  void hide(bool state){
    if(state) _PsystemTurtle_->hide(); 
    else _PsystemTurtle_->show();
  }
  void penDown(bool state){check("penDown"); _PsystemTurtle_->penDown(state);}
  void penUp(bool state){check("penUp"); _PsystemTurtle_->penDown(!state);}
  void left(double dist){check("left"); _PsystemTurtle_->left(dist);}
  void left(int dist){check("left"); left(float(dist));}
  void right(double dist){check("right"); _PsystemTurtle_->right(dist);}
  void right(int dist){check("right"); right(float(dist));}
  void forward(double dist){check("forward"); _PsystemTurtle_->forward(dist);}
  void forward(int dist){check("forward"); forward(float(dist));}

  double sine(double x){return sin(PI*x/180);}
  double cosine(double x){return cos(PI*x/180);}
  double tangent(double x){return tan(PI*x/180);}
  double arcsine(double x){return asin(x)*180/PI;}
  double arccosine(double x){return acos(x)*180/PI;}
  double arctan(double x){return atan(x)*180/PI;}
  double arctan2(double y, double x){return atan2(y,x)*180/PI;}


}


src/canvas.cpp
    #include <simplecpp>
    #include <cstring>
    #include <cstdlib>
    #include <graphics.h>

    using namespace std;
  
    int screen_number=999;
    int screen_width;
    int screen_height;
    int mouse_x,mouse_y; // mouse cordinates
    int ButtonPress,ButtonRelease,MotionNotify,KeyPress;
    char chr;
    int curr_d; // curr_d stores the current screen id
    int screenBG=2; // background screen used to draw 
    int screenTmp=1; // Temp screen used to add sprite(s) at their position
    unsigned int size;
    void *image;  // pointer to screen buffers

    struct LtSprite
    {
      bool operator()(const Sprite* s1, const Sprite* s2) const {
        return s1->getZIndex() < s2->getZIndex();
      }
    };

    namespace canvas{
      double randuv(double u, double v){
        return u + (v-u)*rand()/(1.0+RAND_MAX);
      }

      void srand(unsigned int seed){
        std::srand(seed);
      }

      void abort(){ std::abort(); }

      int canvas_width(){
        return screen_width;
      }
      int canvas_height(){
        return screen_height;
      }

      std::multiset<Sprite *, LtSprite> s_sprite;
      
      /* waiting time before terminating canvas*/
      void wait(float duration){
        delay((int) (duration * 1000));
      }


      int initCanvas(const char window_title[], int width, int height){

        screen_number=initwindow(width,height, window_title, 0,0); // default screen number '0'
        setmousequeuestatus(WM_LBUTTONUP);
        setmousequeuestatus(WM_LBUTTONDOWN);
        setmousequeuestatus(WM_MOUSEMOVE);
        setbkcolor(COLOR(255,255,255));
        cleardevice();  // clears the current screen to white
        setcolor(COLOR(0,0,0));
        setlinestyle(SOLID_LINE, 1, 2);
        if(width == -1){
          
          screen_width = (getwindowwidth( ) -100);
          screen_height = (getwindowheight( ) -100);
        }
        else{
          screen_width = width;
          screen_height= height;
        }

   /* copying current screen to background screen*/
   /* Now background screen is white and ready for drawing*/
   
    	setactivepage(screen_number); 
    	size = imagesize(0,0,screen_width, screen_height);
    	image = malloc(size);
    	getimage(0,0,screen_width, screen_height, image);
    	setactivepage(screenBG);
    	putimage(0, 0, image, COPY_PUT);
    	setcolor(COLOR(0,0,0));
        
    	curr_d=screen_number;

    	free(image);
       		return 0;
      }

      void closeCanvas(){
       
        closegraph( ); 
        s_sprite.clear();    // Delete all elements from set
      }


      void drawLine(XPoint start, XPoint end, Color line_color, unsigned int line_width){
        setactivepage(curr_d);
        setcolor(line_color);
        setlinestyle(SOLID_LINE, 1, line_width);
        line(start.x, start.y, end.x, end.y);
      }

      void imprintLine(short x1, short y1, short x2, short y2, Color line_color, unsigned int line_width){
        setactivepage(curr_d);
        setcolor(line_color);
        setlinestyle(SOLID_LINE, 1, line_width);
        line(x1, y1, x2, y2);
        setcolor(COLOR(0,0,0));
      }

      void drawPoint(XPoint point, Color point_color){ 

        setactivepage(curr_d);
        setcolor(point_color);
        putpixel(point.x, point.y, point_color);
        setcolor(COLOR(0,0,0));
      }

      void drawCircle(XPoint centre, int radius, Color fill_color, bool fill, unsigned int line_width, int line_style){ 
        // Check fill
        if(fill){
          line_width = radius;
          radius /= 2;
        }
        setcolor(fill_color);
        setlinestyle(line_style, 1, line_width);
        arc(centre.x-radius, centre.y - radius, 0, 360, radius);
        setcolor(COLOR(0,0,0));
      }


      void drawEllipse(XPoint centre, int width, int height, Color fill_color, bool fill, unsigned int line_width, int line_style){ 
        setactivepage(curr_d);
        setfillstyle(SOLID_FILL, fill_color);
        setlinestyle(line_style, 1, line_width);
        setcolor(fill_color);
        
        if(fill)
          fillellipse(centre.x , centre.y , width/2, height/2);
        else
          ellipse(centre.x , centre.y , 0, 360, width/2, height/2);
        setcolor(COLOR(0,0,0));
      }

      void drawPolygon(XPoint *points, int npoints, Color fill_color, bool fill, unsigned int line_width, int line_style){ 
        setactivepage(curr_d);
        setfillstyle(SOLID_FILL, fill_color);
        setlinestyle(line_style, 1, line_width);
        setcolor(fill_color);
        if(fill){ 
    	   int poits[npoints*2];
    	   for(int i=0,j=0;i<npoints;i++){
                poits[j++]=points[i].x;
                poits[j++]=points[i].y;
                }
    	        fillpoly(npoints, poits);
        }
        else{
          XPoint pts[npoints + 1];
          int iter;

          // Create array with closed list
          for(iter = 0; iter < npoints; iter ++){
    			pts[iter] = points[iter];
          }
          pts[iter] = points[0];
          //converting 2d array(XPoint type) to 1d array
          int poits[(npoints+1)*2];
    	      for(int i=0,j=0;i<npoints+1;i++)
    	      {
                poits[j++]=pts[i].x;
                poits[j++]=pts[i].y;
                }
    	        drawpoly(npoints+1, poits);
        }
        setcolor(COLOR(0,0,0));
      }

      void addSprite(Sprite *t){
        if(t){
          s_sprite.insert(t);
        }
      }

      void removeSprite(Sprite *t){
        if(t){
          for(std::multiset<Sprite *>::iterator iter = s_sprite.begin();
    	  iter != s_sprite.end(); iter ++){
    	if((*iter) == t){s_sprite.erase(iter); break;}
          }
        }
        repaint();
      }

      void c_imprint(Sprite* s){
        int temp;
        temp = curr_d;
        curr_d= screenBG;
        s->paint();
        curr_d=temp;
      }


      bool globalRepaintFlag = true;
      
      void beginFrame(){
        globalRepaintFlag = false;
      }
      
      void endFrame(){
        globalRepaintFlag = true;
        repaint();
      }


      void repaint(){
        if(screen_number==999){ 
          cout << "You must first call initCanvas before "<<
    	"using any graphics features.\n";
    	wait(5);
          exit(1);
        }

        if(globalRepaintFlag){
          int tl_x=0,tl_y=0,width=canvas_width(),height=canvas_height();
          
            setactivepage(screenBG);
            size = imagesize(0, 0,canvas_width(), canvas_height());
            image = malloc(size);
            getimage(tl_x, tl_y,width, height, image);
            setactivepage(screenTmp);
            putimage(tl_x, tl_y, image, COPY_PUT);

         
            curr_d=screenTmp;
          	for(std::multiset<Sprite *>::iterator iter = s_sprite.begin(); iter != s_sprite.end(); iter ++){
    			(*iter)->paint();
          }

         
        setactivepage(screenTmp);
        getimage(tl_x, tl_y,width, height, image);
        setactivepage(screen_number);
        putimage(tl_x, tl_y, image, COPY_PUT);
        free(image);
        curr_d=screen_number;
        }
      }


      int textWidth(char op){
        string s;
        s= op;
        return  textWidth(s);
      }

      int textWidth(string text){
       
        char txt[text.size() + 1];
        strcpy(txt, text.c_str());
        return textwidth(txt);
      }
      int textHeight(){
       
        return 16; // default height 16
      }
      int textDescent(){
       
        return 0;
      }

      void drawText(Position position, string text, Color clr){
        XPoint p;
        p.x =(int) position.getX(); p.y = (int) position.getY();
        drawText(p, text.c_str(), clr);
      }

      void drawText(float x, float y, string text, Color clr){
        XPoint p;
        p.x = (int)x; p.y = (int)y;
        drawText(p, text.c_str(), clr);
      }


      void drawText(XPoint position, string message, Color clr){
        setactivepage(curr_d);
        char *cstr = new char [message.size()+1];
        strcpy (cstr, message.c_str());
        setcolor(clr);
    	outtextxy(position.x-textWidth(message)/2, position.y -textHeight()/2 - textDescent(), cstr);
        setcolor(COLOR(0,0,0));
      }

      void drawText(XPoint position, const char* text, Color clr){
       
     	setactivepage(curr_d);
        setcolor(clr);
        outtextxy(position.x, position.y, (char*)text);
    	setcolor(COLOR(0,0,0));
      }

      void spriteStatus(){
        cout<<"Count: "<<s_sprite.size()<<endl;
        for(std::multiset<Sprite *>::iterator iter = s_sprite.begin(); iter != s_sprite.end(); iter ++){
          cout<<"["<<"]-->"<<(*iter)<<endl;
        }
      }


        bool Button_Down = false;
        void Up_handler(int x, int y)
        {
            Button_Down = false;
        }
        void nextEvent(XEvent &event)
        {
        //XSync(display, False);
        clearmouseclick(WM_LBUTTONDOWN);
        clearmouseclick(WM_LBUTTONUP);

        while(1)
        {
            ButtonPress=0;
            ButtonRelease=0;
            MotionNotify=0;
            KeyPress =0;
            clearmouseclick(WM_MOUSEMOVE);
            if(ismouseclick(WM_LBUTTONDOWN))
            {
                ButtonPress=1;
                event.type =ButtonPress;
                getmouseclick(WM_LBUTTONDOWN,event.xbutton.x,event.xbutton.y);
                event.xmotion.x=event.xbutton.x;
                event.xmotion.y=event.xbutton.y;
                Button_Down=true; //true
            }
            if(ismouseclick(WM_LBUTTONUP))
            {

                Button_Down=false; //false
                ButtonRelease=2;
                event.type= ButtonRelease;
                getmouseclick(WM_LBUTTONUP,event.xbutton.x,event.xbutton.y);
                event.xmotion.x=event.xbutton.x;
                event.xmotion.y=event.xbutton.y;
            }
            if(Button_Down && ismouseclick(WM_MOUSEMOVE))
                {
                MotionNotify=3;
                event.type= MotionNotify;
                getmouseclick(WM_MOUSEMOVE,event.xbutton.x,event.xbutton.y);
                event.xbutton.x=mousex();
                event.xbutton.y=mousey();
                registermousehandler(WM_LBUTTONUP, Up_handler);
                }

            if(kbhit())
            {
                chr=getch();
                KeyPress=4;
                event.type=KeyPress;
            }

            if(ButtonPress || ButtonRelease || MotionNotify || KeyPress)
                return;
        }
    }

      char charFromEvent(XEvent &event){

            char c;
            if(kbhit())
            {
                c=getch();
                return c;
            }
        return chr;
      }

      int flag=1;
      void echoKey(XEvent &event){

        char c = charFromEvent(event);
        string msg(1,c);
        int tw = textwidth(msg.c_str());
        if(flag==1)
        {
            event.xkey.x=0;
            event.xkey.y=0;
            flag=0;
        }
        drawText(event.xkey.x+tw/2, event.xkey.y, string(1,c));
        event.xkey.x +=tw;
      }


      bool checkEvent(XEvent &event){

            ButtonPress=0;
            ButtonRelease=0;
            MotionNotify=0;
            KeyPress=0;
            clearmouseclick(WM_MOUSEMOVE);
            if(ismouseclick(WM_LBUTTONDOWN))
            {
                ButtonPress=1;
                event.type =ButtonPress;
                getmouseclick(WM_LBUTTONDOWN,event.xbutton.x,event.xbutton.y);
                event.xmotion.x=event.xbutton.x;
                event.xmotion.y=event.xbutton.y;
                Button_Down=true; //true
            }
            if(ismouseclick(WM_LBUTTONUP))
            {

                Button_Down=false; //false
                ButtonRelease=2;
                event.type= ButtonRelease;
                getmouseclick(WM_LBUTTONUP,event.xbutton.x,event.xbutton.y);
                event.xmotion.x=event.xbutton.x;
                event.xmotion.y=event.xbutton.y;
            }
            if(Button_Down && ismouseclick(WM_MOUSEMOVE))
                {
                MotionNotify=3;
                event.type= MotionNotify;
                getmouseclick(WM_MOUSEMOVE,event.xbutton.x,event.xbutton.y);
                event.xbutton.x=mousex();
                event.xbutton.y=mousey();
                registermousehandler(WM_LBUTTONUP, Up_handler);
                }
            if(kbhit())
            {
                KeyPress=4;
                event.type=KeyPress;
                chr=getch();
            }
       if(ButtonPress || ButtonRelease || MotionNotify || KeyPress)
        return true;
       return false;
      }

      bool mouseDragEvent(XEvent &event){
        return event.type==MotionNotify;
      }

      bool mouseButtonPressEvent(XEvent &event){
        return event.type==ButtonPress;
      }


      bool mouseButtonReleaseEvent(XEvent &event){
        return event.type==ButtonRelease;
      }

      bool keyPressEvent(XEvent &event){

        return event.type == KeyPress;
      }

      int getClick(){

        XEvent event;
      
        clearmouseclick(WM_LBUTTONDOWN);
        clearmouseclick(WM_LBUTTONUP);
        clearmouseclick(WM_MOUSEMOVE);
        while(true){
          nextEvent(event);
          if(ButtonPress){
            return event.xbutton.x * 65536 + event.xbutton.y;
          }

        }
        return 0;
      }

    }
    Color COLOR(const char *color_string){
   
    if(screen_number==999){ 
          cout << "You must first call initCanvas before "<<
    	"using any graphics features.\n";
    	wait(5);
          exit(1);
        }

          if( strcmp(color_string, "red")==0)
              return ::COLOR(255,0,0);
          else if( strcmp(color_string, "green")==0)
              return ::COLOR(0,255,0);
          else if( strcmp(color_string, "blue")==0)
              return ::COLOR(0,0,255);
          else if( strcmp(color_string, "yellow")==0)
              return ::COLOR(255,255,0);
          else if( strcmp(color_string, "black")==0)
              return ::COLOR(0,0,0);
          else
              return ::COLOR(255,255,255);
   
      }

